### *Fadlila Agustina*
### *21083010050*
### *Machine Learning I (C)*

### Menampilkan Dataset 

In [1]:
# mengimpor pandas sbg pd dan numpy sbg np yg digunakan dalam pengolahan data
import pandas as pd
import numpy as np

# Nama_variabel diinisialisasikan sbg list yg berisi nama-nama kolom
Nama_Variabel=['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
# fungsi pd.read_csv digunakan utk membaca file sbg objek DataFrame
# parameter header=None digunakan  utk menunjukkan bahwa data tdk memiliki baris header 
# delim_whitespace=True menunjukkan bahwa delimiter yg digunakan adalah spasi
# parameter names digunakan utk menentukan nama kolom dr dataset yg akan dimuat ke dlm DataFrame dan disesuaikan dgn var Nama_Variabel
df = pd.read_csv('housing.csv', header = None, delim_whitespace=True, names=Nama_Variabel)

# utk menampilkan DataFrame
print(df)

        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14

In [53]:
# fungsi yg digunakan utk menampilkan 5 baris utama dari DataFrame
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

### Tahapan Pre-Processing Data

In [54]:
# fungsi yg digunakan utk menampilkan dimensi dr DataFrame dan akan menunjukkan jumlah baris dan kolom dr DataFrame
np.shape(df)

(506, 14)

In [55]:
# fungsi yg digunakan utk menampilkan tipe data dr setiap kolom
df.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS         int64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

In [56]:
# mengubah tipe data pada kolom CHAS yg awalnya int64 menjadi tipe data categori
# kategori adalah salah satu tipe data pandas yg digunakan utk menyimpan data dgn nilai diskrit (terbatas)
df["CHAS"] = df["CHAS"].astype('category')

# fungsi yg digunakan utk menampilkan tipe data dr setiap kolom
df.dtypes

CRIM        float64
ZN          float64
INDUS       float64
CHAS       category
NOX         float64
RM          float64
AGE         float64
DIS         float64
RAD           int64
TAX         float64
PTRATIO     float64
B           float64
LSTAT       float64
MEDV        float64
dtype: object

In [57]:
# fungsi yg digunakan utk menampilkan ringkasan statistik deskriptif dr setiap kolom
df.describe()

CRIM          ZN       INDUS         NOX          RM         AGE  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.554695    6.284634   68.574901   
std      8.601545   23.322453    6.860353    0.115878    0.702617   28.148861   
min      0.006320    0.000000    0.460000    0.385000    3.561000    2.900000   
25%      0.082045    0.000000    5.190000    0.449000    5.885500   45.025000   
50%      0.256510    0.000000    9.690000    0.538000    6.208500   77.500000   
75%      3.677083   12.500000   18.100000    0.624000    6.623500   94.075000   
max     88.976200  100.000000   27.740000    0.871000    8.780000  100.000000   

              DIS         RAD         TAX     PTRATIO           B       LSTAT  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.795043    9.549407  408.237154   18.455534  356.674032   12.653063   
std      2.105710    8.707259  168.537116    2.164946   91.294864    7.141062   
min      1.129600    1.000000  187.000000   12.600000    0.320000    1.730000   
25%      2.100175    4.000000  279.000000   17.400000  375.377500    6.950000   
50%      3.207450    5.000000  330.000000   19.050000  391.440000   11.360000   
75%      5.188425   24.000000  666.000000   20.200000  396.225000   16.955000   
max     12.126500   24.000000  711.000000   22.000000  396.900000   37.970000   

             MEDV  
count  506.000000  
mean    22.532806  
std      9.197104  
min      5.000000  
25%     17.025000  
50%     21.200000  
75%     25.000000  
max     50.000000

In [58]:
# fungsi yg digunakan utk menghitung jumlah data pada setiap kategori pada kolom CHAS karena sudah diubah menjadi tipe data kategori
df['CHAS'].value_counts()

0    471
1     35
Name: CHAS, dtype: int64

In [59]:
# fungsi yg digunakan utk mengecek apakah terdapat nilai kosong (missing value) pada tiap kolom
# jika True, maka kolom tsb mengandung setidaknya satu nilai kosong
# jika False, maka kolom tsb tdk mengandung nilai kosong
df.isnull().any()

CRIM       False
ZN         False
INDUS      False
CHAS       False
NOX        False
RM         False
AGE        False
DIS        False
RAD        False
TAX        False
PTRATIO    False
B          False
LSTAT      False
MEDV       False
dtype: bool

In [60]:
# mengimport modul preprocessing dr library scikit-learn
# preprocessing adalah modul yg berisi berbagai macam fungsi utk melakukan preprocessing pada data
from sklearn import preprocessing
# membuat objek MinMaxScaler() dr kelas preprocessing.MinMaxScaler() yg digunakan utk melakukan transformasi data dgn metode scaling minimum-maximum
transformasi_min_max = preprocessing.MinMaxScaler()

### Pemodelan Data dengan Regresi

In [61]:
# menginisialisasikan var target dgn nilai string MEDV dan digunakan utk menyimpan nama kolom yg menjadi target atau var dependen dlm datase
target = 'MEDV'
# menghapus kolom MEDV menggunakan metode drop
# axis = 1 digunakan utk menandakan bahwa penghapusan dilakukan pada kolom
x = df.drop('MEDV', axis =1)
# pengambilan kolom MEDV dr dataset dan dimasukkan ke dlm var y
y = df[target]
# menginisialisasikan var varx dgn list yg berisi nama kolom pada dataset dan list ini digunakan sbg nama kolom pada dataset x
varx = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT']
# normalisasi data menggunakan metode MinMaxScaler() dr library sklearn.preprocessing
# data yg dinormalisasikan dimasukkan ke dlm DataFrame baru dgn nama kolom yg ditentukan pada varx
x = pd.DataFrame(data = transformasi_min_max.fit_transform(x), columns = varx)

In [62]:
# mengimpor modul train_test_split dari library sklearn.model_sleection yg digunakan utk membagi dataset menjadi data training dan data testing
from sklearn.model_selection import train_test_split
# melakukan pemisahan dataset menjadi 4 bagian (x_train, x_test, y_train, y_test)
# var x merupakan data independen yg dipakai utk memprediksi var target y
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.80, random_state = 1)

In [63]:
# mengimpor modul LinearRegression dr library sklearn.linear_modal yg digunakan utk membuat model regresi linear
from sklearn.linear_model import LinearRegression
# mengimpor modul mean_squared_error dan r2_score dr library sklearn.metrics yg digunakan utk mengukur kinerja model
from sklearn.metrics import mean_squared_error, r2_score

# membuat objek model regresi linear yg disimpan dlm var linear_model
linear_model = LinearRegression()
# melatih model regresi linear dgn memasukkan data latih x dan y ke dlm model menggunakan metode fit()
linear_model.fit(x_train, y_train)

LinearRegression()

In [71]:
# mengimpor modul PolynomialFeatures dr library sklearn.preprocessing yg digunakan utk membuat fitur polinomial
from sklearn.preprocessing import PolynomialFeatures
# membuat objek PolynomialFeatures dgn prameter degree 2 yg menunjukkan bahwa fitur polinomial yg dibuat adalah dr derajat 0 sampai 2
polynomial_features = PolynomialFeatures(degree = 2)
# mentransformasikan x_train menjadi fitur polinomial dgn menggunakan objek polynomial_features di atas dan disimpan ke dlm var x_poly_train
x_poly_train = polynomial_features.fit_transform(x_train)
# membuat objek model regresi linear yg disimpan dlm var poly_model
poly_model = LinearRegression()
# melatih model regresi linear dgn memasukkan x_poly_train dan y_train ke dlm model menggunakan metode fit()
poly_model.fit(x_poly_train, y_train)
# mentransformasikan x_test menjadi fitur polinomial menggunakan objek polynomial_features dan disimpan dlm var x_poly_test
x_poly_test = polynomial_features.fit_transform(x_test)

In [72]:
# mengimpor modul ridge dr library sklearn.linear_model yg digunakan utk membuat model regresi ridge
from sklearn.linear_model import Ridge

# membuat objek model regresi ridge dgn alpha = 1
# alpha adalah hyperparameter yg mengontrol kompleksitas model, semakin besar nilai alpha, maka semakin besar pembatasan pada koef model dan semakin sederhana modelnya
model_ridge = Ridge(alpha = 1)
# melatih model regresi ridge dgn memasukkan x_train dan y_train ke dlm model menggunakan metode fit()
# algoritma optimasi akan menyesuaikan koef utk meminimalkan jumlah kuadrat residual antara nilai prediksi dan nilai target
model_ridge.fit(x_train, y_train)

Ridge(alpha=1)

In [73]:
# mengimpor modul lasso dr library sklearn.linear_model yg digunakan utk membuat model regresi lasso
from sklearn.linear_model import Lasso

# membuat objek model regresi lasso dgn alpha = 1
# alpha adalah hyperparameter yg mengontrol kompleksitas model, semakin besar nilai alpha, maka semakin besar pembatasan pada koef model dan semakin sederhana modelnya
model_lasso = Lasso(alpha = 1)
# melatih model regresi lasso dgn memasukkan x_train dan y_train ke dlm model menggunakan metode fit()
# algoritma optimasi akan menyesuaikan koef utk meminimalkan jumlah absolut residual antara nilai prediksi dan nilai target dgn mempertahankan koef yg signifikan dan menghilangkan yg tdk signifikan
model_lasso.fit(x_train, y_train)

Lasso(alpha=1)

### Evaluasi Data Testing dengan Model Regresi Linear

In [74]:
# melakukan prediksi pada data x_train menggunakan linear_model yg sudah dilatih sblmnya, hasil prediksi akan disimpan ke dlm var pred_y_train
pred_y_train = linear_model.predict(x_train)
# menghitung rmse utk data training, rmse adalah metrik evaluasi utk mengukur sbrp dekat prediksi model dgn nilai sebenarnya
# semakin kecil nilai rmse, maka semakin baik performa modelnya
# menggunakan fungsi mean_squared_error() utk menghtung nilai mse kemudian mengambil akar kuadratnya menggunakan fungsi np.sqrt()
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train)))
# menghitung nilai r2 utk data training menggunakan fungsi r2_score
# r2 adalah metrik evaluasi yg mengukur sbrp baik model dpt menjelaskan variasi data
# nilai r2 berkisar antara 0 dan 1, semakin besar nilai r2, maka menunjukkan semakin baik performa modelnya
r2 = r2_score(y_train, pred_y_train)

# melakukan prediksi pada x_test menggunakan linear_model yg sudah dilatih sblmnya, hasil prediksi disimpan dlm var pred_y_test
pred_y_test = linear_model.predict(x_test)
# menghitung nilai rmse utk data testing
# menggunakan fungsi mean_squared_error() utk menghitung nilai mse kemudian mengambil akar kuadratnya menggunakan fungsi np.sqrt()
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test)))
# menghitung nilai r2 utk data testing menggunakan fungsi r2_score
r2_2 = r2_score(y_test, pred_y_test)

print("--------------------------------------")
print("PERFORMA MODEL UNTUK DATA TRAINING :")
print("--------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data training menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse)) 
# mencetak string yg dihasilkan dr nilai r2 data training menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2))
print("\n--------------------------------------")
print("PERFORMA MODEL UNTUK DATA TESTING :")
print("--------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse_2))
# # mencetak string yg dihasilkan dr nilai r2 data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2_2))

--------------------------------------
PERFORMA MODEL UNTUK DATA TRAINING :
--------------------------------------
NILAI RMSE -> 4.67053618630709
NILAI R2 -> 0.7528990263438373

--------------------------------------
PERFORMA MODEL UNTUK DATA TESTING :
--------------------------------------
NILAI RMSE -> 4.814941555935469
NILAI R2 -> 0.7219648298245814


In [75]:
# memprediksi nilai y_train berdasarkan poly_model yg sudah dilatih sblmnya dgn memasukkan nilai x_poly_train dan hasil prediksi disimpan ke dlm var pred_y_train_poly
pred_y_train_poly = poly_model.predict(x_poly_train)
# menghitung rmse dr model polynomial yg sudah dilatih dgn membandingkan y_train dgn pred_y_train_poly menggunakan fungsi mean_squared_error()
# fungsi np.sqrt digunakan utk menghitung nilai akar dr mse
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_poly)))
# menghitung nilai r2 dr model polynomial yg sudah dilatih dgn membandingkan y_train dgn pred_y_train_poly menggunakan fungsi r2_score
r2 = r2_score(y_train, pred_y_train_poly)

# memprediksi y_test berdasarkan poly_model dgn memasukkan x_poly_test dan hasilnya akan disimpan dlm var pred_y_test_poly
pred_y_test_poly = poly_model.predict(x_poly_test)
# menghitung nilai rmse dgn membandingkan y_test dgn pred_y_test_poly menggunakan fungsi mean_squared_error()
# fungsi np.sqrt digunakan utk menghitung nilai akar dr nilai mse
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_poly)))
# menghitung nilai r2 dgn membanding y_test dgn pred_y_test_poly menggunakan fungsi r2
r2_2 = r2_score(y_test, pred_y_test_poly)

print("-----------------------------------------------------------")
print("PERFORMA MODEL POLINOMIAL DERAJAT 2 UNTUK DATA TRAINING :")
print("-----------------------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data training menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse))
# mencetak string yg dihasilkan dr nilai r2 data training menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2))
print("\n-----------------------------------------------------------")
print("PERFORMA MODEL POLINOMIAL DERAJAT 2 UNTUK DATA TESTING :")
print("-----------------------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse_2))
# mencetak string yg dihasilkan dr nilai r2 data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2_2))

-----------------------------------------------------------
PERFORMA MODEL POLINOMIAL DERAJAT 2 UNTUK DATA TRAINING :
-----------------------------------------------------------
NILAI RMSE -> 0.3988691356372389
NILAI R2 -> 0.9981978047601218

-----------------------------------------------------------
PERFORMA MODEL POLINOMIAL DERAJAT 2 UNTUK DATA TESTING :
-----------------------------------------------------------
NILAI RMSE -> 122.16834247851112
NILAI R2 -> -177.99259272908384


In [76]:
# memprediksi y_train berdasarkan model_ridge dgn memasukkan x_train dan hasil prediksi disimpan dlm var pred_y_train_ridge
pred_y_train_ridge = model_ridge.predict(x_train)
# menghitung rmse dgn membandingkan y_train dgn pred_y_train_ridge menggunakan fungsi mean_squared_error
# np.sqrt digunakan utk menghitung nilai akar dr mse
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_ridge)))
# menghitung r2 dgn membandingkan y_train dgn pred_y_train_ridge menggunakan fungsi r2_score
r2 = r2_score(y_train, pred_y_train_ridge)

# memprediksi y_test berdasarkan model_ridge dgn memasukkan x_test dan hasil prediksi disimpan dlm var pred_y_test_ridge
pred_y_test_ridge = model_ridge.predict(x_test)
# menghitung rmse dgn membandingkan y_test dgn pred_y_test_ridge dan hasil prediksi disimpan dlm var pred_y_test_ridge
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_ridge)))
# menghitung r2 dgn membandingkan y_test dgn pred_y_test_ridge menggunakan fungsi r2_score
r2_2 = r2_score(y_test, pred_y_test_ridge)

print("------------------------------------------")
print("PERFORMA MODEL RIDGE UNTUK DATA TRAINING :")
print("------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data training menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse))
# mencetak string yg dihasilkan dr nilai r2 data training menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2))
print("\n------------------------------------------")
print("PERFORMA MODEL RIDGE UNTUK DATA TESTING :")
print("------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse_2))
# mencetak string yg dihasilkan dr nilai r2 data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2_2))

------------------------------------------
PERFORMA MODEL RIDGE UNTUK DATA TRAINING :
------------------------------------------
NILAI RMSE -> 4.963050499300717
NILAI R2 -> 0.7209780544959115

------------------------------------------
PERFORMA MODEL RIDGE UNTUK DATA TESTING :
------------------------------------------
NILAI RMSE -> 4.988416251758976
NILAI R2 -> 0.701569596206263


In [77]:
# memprediksi y_train berdasarkan model_lasso dgn memasukkan x_train dan hasil prediksi disimpan dlm var pred_y_train_lasso
pred_y_train_lasso = model_lasso.predict(x_train)
# menghitung rmse dgn membandingkan y_train dgn pred_y_train_lasso menggunakan fungsi mean_squared_error()
# fungsi np.sqrt digunakan utk menghitung nilai akar dr mse
rmse = (np.sqrt(mean_squared_error(y_train, pred_y_train_lasso)))
# menghitung r2 dgn membandingkan y_train dgn pred_y_train_lasso menggunakan fungsi r2
r2 = r2_score(y_train, pred_y_train_lasso)

# memprediksi y_test berdasarkan model_lasso dgn memasukkan x_test dan hasil prediksi disimpan dlm var pred_y_test_lasso
pred_y_test_lasso = model_lasso.predict(x_test)
# menghitung rmse dgn membandingkan y_test dgn pred_y_test_lasso menggunakan fungsi mean_squared_error()
# fungsi np.sqrt digunakan utk menghitung nilai akar dr mse
rmse_2 = (np.sqrt(mean_squared_error(y_test, pred_y_test_lasso)))
# menghitung r2 dgn membandingkan y_test dgn pred_y_test_lasso menggunakan fungsi r2_score
r2_2 = r2_score(y_test, pred_y_test_lasso)

print("------------------------------------------")
print("PERFORMA MODEL LASSO UNTUK DATA TRAINING :")
print("------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data training menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse))
# mencetak string yg dihasilkan dr nilai r2 data training menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2))
print("\n------------------------------------------")
print("PERFORMA MODEL LASSO UNTUK DATA TESTING :")
print("------------------------------------------")
# mencetak string yg dihasilkan dr nilai rmse data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai rmse ke dlm string tsb
print("NILAI RMSE -> {}".format(rmse_2))
# mencetak string yg dihasilkan dr nilai r2 data testing menggunakan fungsi format() yg digunakan utk memasukkan nilai r2 ke dlm string tsb
print("NILAI R2 -> {}".format(r2_2))

------------------------------------------
PERFORMA MODEL LASSO UNTUK DATA TRAINING :
------------------------------------------
NILAI RMSE -> 7.9731339072884255
NILAI R2 -> 0.27988958819908916

------------------------------------------
PERFORMA MODEL LASSO UNTUK DATA TESTING :
------------------------------------------
NILAI RMSE -> 7.77037539712362
NILAI R2 -> 0.2758947488538893


### Kesimpulan

R-square dan RMSE adalah dua metrik evaluasi yang umum digunakan untuk mengevaluasi kualitas model regresi. R-square mengukur seberapa besar proporsi variabilitas dalam data yang dapat dijelaskan oleh model regresi yang dihasilkan, sementara RMSE mengukur seberapa besar kesalahan prediksi model dalam satuan yang sama dengan data yang diamati. Sehingga, model regresi yang baik adalah model yang memberikan nilai R-square yang tinggi dan nilai RMSE yang rendah.

Dalam hal ini, Model Linear dipilih sebagai model terbaik karena memiliki nilai R-Square sebesar 0.7219648298245814 dan nilai RMSE sebesar 4.814941555935469. Hal ini menunjukkan bahwa model memiliki kemampuan yang baik dalam menjelaskan variasi data yang diamati, dan RMSE yang rendah menunjukkan bahwa model memberikan prediksi yang akurat dan dekat dengan nilai sebenarnya. Namun, perlu diingat bahwa pemilihan model terbaik tidak hanya bergantung pada R-square dan RMSE saja, namun juga harus mempertimbangkan faktor lain seperti akurasi, kecepatan, dan kompleksitas model.